File: Neetcode Questions/Trees/diameter-of-binary-tree.cpp

// https://leetcode.com/problems/diameter-of-binary-tree/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */


class Solution {
public:

    int height(TreeNode*root){
        if(!root) return 0;
        return 1 + max(height(root->left), height(root->right));
    }

    int diameterOfBinaryTree(TreeNode* root) {
        if(!root) return 0;
        int h = height(root->left) + height(root->right);
        int ld = diameterOfBinaryTree(root->left);
        int rd = diameterOfBinaryTree(root->right);
        return max({h, ld, rd}); 
    }
};

File: Neetcode-150-js/Trees/diameter-of-binary-tree.js

// https://leetcode.com/problems/diameter-of-binary-tree/description/

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */

var height = function(root){
    if(!root) return 0;
    return 1 + Math.max(height(root.left), height(root.right));
}

var diameterOfBinaryTree = function(root) {
    if(!root) return 0;
    let h = height(root.left) + height(root.right);
    let ld = diameterOfBinaryTree(root.left);
    let rd = diameterOfBinaryTree(root.right);
    return Math.max(h, ld, rd);
};